File: entities/entity.ts
import { Color } from "../primitives/color";
import { Ray } from "../primitives/ray";
import { RayHit } from "../primitives/ray-hit";
import { Transform } from "../primitives/transform";

export abstract class Entity {
	constructor (
		public color: Color,
		public transform: Transform
	) {}

	abstract intersect(ray: Ray): RayHit | null;
}

File: src/entities/entity.rs
use crate::primitives::{Aabb, Ray, RayHit};

pub trait Entity: Sync + Send {
	fn intersect(&self, ray: &Ray) -> Option<RayHit>;

	/// returns an axis-aligned bounding box that fully encloses this entity.
	fn bounding_box(&self) -> Aabb;
}
